### PR TITLE
Add support flash for write protection and readout protection for STM32F4

### PR DESCRIPTION
**Pull request by duda-patryk**
This PR introduces support for STM32F4 flash write protection (persistent across reboots) and readout protection (temporary or permanent).

The flash API was extended by functions responsible for checking and changing WP and RDP state. Enabling RDP permanently is guarded by FLASH_ALLOW_PERMANENT_READOUT_PROTECTION option. Disabling RDP (which triggers mass erase) is guarded by FLASH_ALLOW_DISABLING_READOUT_PROTECTION.

I guess that the most controversial part is API change, so any remarks are welcome!

Best regards,
Patryk

### CODE REVIEW

**Review by erwango — CHANGES_REQUESTED** [review]
Some preliminary requests about API introduction.

File: drivers/flash/Kconfig
Comment (semihalf-duda-patryk) [reply drivers/flash/Kconfig]: Done

**Comment by semihalf-duda-patryk** [comment]
Make `FLASH_WRITE_PROTECT_API` and `FLASH_READOUT_PROTECTION_API` disabled by default. Add `__unused` to `flash_stm32_option_bytes_lock`.

File: drivers/flash/Kconfig
Comment (semihalf-duda-patryk) [reply drivers/flash/Kconfig]: Done.

File: drivers/flash/Kconfig
Comment (semihalf-duda-patryk) [reply drivers/flash/Kconfig:43]: This is information if WP or RDP is supported by flash driver (similar to `FLASH_HAS_PAGE_LAYOUT`) rather that hardware description. IIRC all STM32 devices support WP and RDP.

**Comment by erwango** [comment]
Since this PR introduces changes to Flash API (which I guess is a stable API), you need first to get the API change approved. This would be faster to do so by splitting API changes in a dedicated PR.

**Comment by semihalf-duda-patryk** [comment]
PR for flash API changes https://github.com/zephyrproject-rtos/zephyr/pull/53441

File: drivers/flash/Kconfig
Comment (erwango) [reply drivers/flash/Kconfig:43]: ok, fine to me.

**Comment by semihalf-duda-patryk** [comment]
Modified implementation of WP and RDP to work with "extended operations" API (PR #53441).

**Comment by semihalf-duda-patryk** [comment]
Fixed test issues.

**Comment by semihalf-duda-patryk** [comment]
Added extra operations for blocking control and option registers until reboot.

API change #53441 was merged, so we go back to this review now.

**Review by erwango — CHANGES_REQUESTED** [review]
Some minor comments.
I only could have a quick review for now. Looks ok, but I didn't had time to get deep into details.

What would be great is to add some test (or at least sample) to compile and demonstrate at least some of the API (I understand that some use cases are hard to demonstrate)

File: include/zephyr/drivers/flash/st_flash_api_extensions.h
Comment (de-nordic) [inline include/zephyr/drivers/flash/st_flash_api_extensions.h:39]: Do we want to have EX_OP definitions Kconfig dependent?
Shouldn't we also have all defines grouped together to allow easier to enum existing and add new?

File: drivers/flash/CMakeLists.txt
Comment (semihalf-duda-patryk) [reply drivers/flash/CMakeLists.txt]: No. I made compiling the file to depend on `CONFIG_FLASH_EX_OP_ENABLED`

File: dts/bindings/flash_controller/st,stm32-flash-controller.yaml
Comment (semihalf-duda-patryk) [reply dts/bindings/flash_controller/st,stm32-flash-controller.yaml:17]: Done. Thanks!

File: include/zephyr/drivers/flash/st_flash_api_extensions.h
Comment (semihalf-duda-patryk) [reply include/zephyr/drivers/flash/st_flash_api_extensions.h]: Moved definitions to the enum.

File: include/zephyr/drivers/flash/st_flash_api_extensions.h
Comment (semihalf-duda-patryk) [reply include/zephyr/drivers/flash/st_flash_api_extensions.h:1]: Done.

**Comment by semihalf-duda-patryk** [comment]
Implemented tests for write protection and reading readout protection status.

Implementing test for changing readout protection is hard to implement, because once RDP is enabled, we can't disable it without mass erase. Similar for disabling access to control/option register - reset is needed to remove protection.

File: drivers/flash/Kconfig.stm32
Comment (de-nordic) [inline drivers/flash/Kconfig.stm32:17]: Not an STM32 expert, but shouldn't the `FLASH_HAS_EX_OP` be selected by `SOC_SERIES_STM32F4X` and other SoC specific configs, where the SoC drivers actually provide support for any `ex_op`?

File: drivers/flash/Kconfig.stm32
Comment (de-nordic) [inline drivers/flash/Kconfig.stm32:32]: Try naming Kconfigs in some structured way, for example here you have `FLASH_STM32_WRITE_PROTECT`  and the `FLASH_STM32_BLOCK_WRITE_PROTECT_DISABLING` is additional option to `FLASH_STM32_WRITE_PROTECT`. `FLASH_STM32_WRITE_PROTECT_DISABLE_PREVENTION` (sorry could not get myself to write BLOCK), would be more intuitive.

File: drivers/flash/flash_stm32.h
Comment (de-nordic) [inline drivers/flash/flash_stm32.h:265]: Unneeded empty line.

File: include/zephyr/drivers/flash/stm32_flash_api_extensions.h
Comment (de-nordic) [inline include/zephyr/drivers/flash/stm32_flash_api_extensions.h:23]: Just a note, no change required: the header may provide `flash_stm32_ex_op_sector_wp_get()` `flash_stm32_ex_op_sector_wp_set()` inline functions in this header that will be passing reduced set of parameters to flash_ex_op, users may appreciate. 
This would probably work for all the others `FLASH_STM32_EX` commands provided.

**Review by de-nordic — COMMENTED** [review]
Looks pretty.
Some minor comments SMT32 maintainer may disagree with.

File: drivers/flash/Kconfig.stm32
Comment (erwango) [reply drivers/flash/Kconfig.stm32:17]: Agreed.

Also, I wonder if it would be acceptable to select `FLASH_HAS_EX_OP` based on a device tree input (new compatible/new property). I know this was not acceptable some time ago (SW/HW separation stuff), but maybe this policy has evolved.

File: drivers/flash/flash_stm32_ex_op.c
Comment (semihalf-duda-patryk) [reply drivers/flash/flash_stm32_ex_op.c]: Fixed.

File: drivers/flash/flash_stm32.h
Comment (semihalf-duda-patryk) [reply drivers/flash/flash_stm32.h:259]: Fixed.

File: drivers/flash/Kconfig.stm32
Comment (de-nordic) [reply drivers/flash/Kconfig.stm32:17]: Like `compatible = "stm32,flash,sector_wp"`?

Comment (erwango) [reply drivers/flash/Kconfig.stm32:17]: Should be `st,stm32-flash-something`

Comment (semihalf-duda-patryk) [reply drivers/flash/Kconfig.stm32:17]: Should I add `st,stm32-flash-ex-op` to existing flash compatible in `dts/arm/st/f4/stm32f4.dtsi` and check it in Kconfig?

Comment (de-nordic) [reply drivers/flash/Kconfig.stm32:17]: I am not sure how this would work if we have to add more compatibles per device/node to indicate supported features.

Comment (semihalf-duda-patryk) [reply drivers/flash/Kconfig.stm32:17]: Are we talking about compatibility in DTS which indicates that *any* extended operation is supported or about indicating which extended operations are supported?

Support for extended operations (`flash_ex_op()` syscall) can be enabled on all devices supported by the driver (though most devices don't support any extended operation), but WP, RDP, etc. is currently supported on STM32F4 series.

Comment (erwango) [reply drivers/flash/Kconfig.stm32:17]: I was talking about any extended operation. But I agree this can make things complex if we need to enter details about which operations are supported if not all of them are added at once or if for some reason one is not supported.

So, coming back to the initial point, `select FLASH_HAS_EX_OP` is valid on all series even if no OP is implemented. Is that correct  ?

Comment (semihalf-duda-patryk) [reply drivers/flash/Kconfig.stm32:17]: > So, coming back to the initial point, `select FLASH_HAS_EX_OP` is valid on all series even if no OP is implemented. Is that correct ?

Yes. `flash_stm32_ex_op()` will return `-ENOTSUP` if requested OP is not supported.

Comment (de-nordic) [reply drivers/flash/Kconfig.stm32:17]: I think that having `select FLASH_HAS_EX_OP` if your driver, or your device, does not actually provide implementation invalidates entire purpose of the Kconfig. The option directly states "this thing has that feature".

Comment (semihalf-duda-patryk) [reply drivers/flash/Kconfig.stm32:17]: I can make it `select FLASH_HAS_EX_OP if SOC_SERIES_STM32F4X`

Comment (semihalf-duda-patryk) [reply drivers/flash/Kconfig.stm32:17]: > Yeah, but your implementation of `flash_ex_op` can do exactly the same without calling the driver: when none of drivers selects `FLASH_HAS_EX_OP` then the code that goes for flash driver will not even be enabled, shortening call paths to just returning `-ENOTSUP` directly from `FLASH_HAS_EX_OP`.

That's right, but @erwango asked if it's okay to `select FLASH_HAS_EX_OP` even if no extended operations is implemented.

Comment (de-nordic) [reply drivers/flash/Kconfig.stm32:17]: > 
> That's right, but @erwango asked if it's okay to `select FLASH_HAS_EX_OP` even if no extended operations are implemented.

IMHO no. Why? For example you have additional part of software that is common for all devices but does additional work when `FLASH_HAS_EX_OP` is enabled (for some reason, comment is for general usage of _HAS_ Kconfig options), now you will have that part of code compiled in, and probably executed, even though your driver does not require it to run but it selects `FLASH_HAS_EX_OP` for no reason.

Comment (semihalf-duda-patryk) [reply drivers/flash/Kconfig.stm32:17]: I agree that we should avoid compiling and executing code that effectively does nothing. I thought that the question was about potential problems when `FLASH_HAS_EX_OP` is selected, but no OP is implemented.

Anyway, selecting `FLASH_HAS_EX_OP` only on devices that have any OP implemented is a good thing. The question is if it's okay to do `select FLASH_HAS_EX_OP if SOC_SERIES_STM32F4X` or we should go with DTS solution.

Comment (erwango) [reply drivers/flash/Kconfig.stm32:17]: Let's go with t's okay to do `select FLASH_HAS_EX_OP if SOC_SERIES_STM32F4X` for now. We can tune things later

File: drivers/flash/flash_stm32.c
Comment (semihalf-duda-patryk) [reply drivers/flash/flash_stm32.c:338]: @erwango should I change it to `else if` ?

Comment (erwango) [reply drivers/flash/flash_stm32.c:338]: that would be nicer indeed.

Comment (semihalf-duda-patryk) [reply drivers/flash/flash_stm32.c]: Done.

File: tests/drivers/flash/stm32/src/main.c
Comment (semihalf-duda-patryk) [reply tests/drivers/flash/stm32/src/main.c]: Done.

**Review by de-nordic — DISMISSED** [review]
Looks OK. Thanks!

**Review by erwango — DISMISSED** [review]
Great work, thanks

**Comment by semihalf-duda-patryk** [comment]
Created #56181 for `drivers.flash.common.default` unit test build failure.

File: drivers/flash/Kconfig.stm32
Comment (mbolivar-nordic) [reply drivers/flash/Kconfig.stm32]: I did not read all of this discussion, but the official policy on software specific features in DTS is documented here (see "the zephyr, prefix"):

https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#the-zephyr-prefix

Comment (erwango) [reply drivers/flash/Kconfig.stm32]: Clear, thanks @mbolivar-nordic

**Comment by semihalf-duda-patryk** [comment]
Patches were rebased and the tests are passing now. @de-nordic, @erwango  could you approve it once again?

**Review by erwango — APPROVED** [review]

**Review by de-nordic — APPROVED** [review]

**Pull request merged**
